packetbeat/protos/smtp/trans.go

```diff
@@ -0,0 +1,128 @@
+package smtp
+
+import (
+	"time"
+
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/logp"
+	"github.com/elastic/beats/packetbeat/procs"
+	"github.com/elastic/beats/packetbeat/protos/applayer"
+)
+
+type transactions struct {
+	config *transactionConfig
+
+	requests  messageList
+	responses messageList
+
+	onTransaction transactionHandler
+}
+
+type transactionConfig struct {
+	transactionTimeout time.Duration
+}
+
+type transactionHandler func(requ, resp *message) error
+
+// List of messages available for correlation
+type messageList struct {
+	head, tail *message
+}
+
+func (trans *transactions) init(c *transactionConfig, cb transactionHandler) {
+	trans.config = c
+	trans.onTransaction = cb
+}
+
+func (trans *transactions) onMessage(
+	tuple *common.IPPortTuple,
+	dir uint8,
+	msg *message,
+) error {
+	msg.Tuple = *tuple
+	msg.Transport = applayer.TransportTCP
+	msg.CmdlineTuple = procs.ProcWatcher.FindProcessesTuple(&msg.Tuple)
+
+	if msg.IsRequest {
+		if isDebug {
+			debugf("Received request with tuple: %s", tuple)
+		}
+		trans.requests.append(msg)
+	} else {
+		if isDebug {
+			debugf("Received response with tuple: %s", tuple)
+		}
+		trans.responses.append(msg)
+	}
+
+	return trans.correlate()
+}
+
+func (trans *transactions) correlate() error {
+	requests := &trans.requests
+	responses := &trans.responses
+
+	// drop responses with missing requests, unless they are "prompt"
+	// responses (220, 421, etc.)
+	if requests.empty() {
+		for !responses.empty() {
+			resp := responses.pop()
+			switch resp.statusCode {
+			case 220, 421:
```

Comment:
what about 554? Any more codes that are commonly used when prompting?

Reply:
I'll check again to make sure, but 554 can only happen as a response to `DATA`.
I'm not aware of any other prompt response codes.